File: Java/src/com/leetcode/Medium/AsteroidCollision.java
package com.leetcode.Medium;

import java.util.*;

// 735. Asteroid Collision
public class AsteroidCollision {
    public int[] asteroidCollision(int[] asteroids) {
        LinkedList<Integer> stack = new LinkedList<>();
        for (int i = 0; i < asteroids.length; i++) {
            if (asteroids[i] > 0 || stack.isEmpty() || stack.getLast() < 0)
                stack.add(asteroids[i]);
            else if (stack.getLast() <= -asteroids[i])
                if (stack.pollLast() < -asteroids[i])
                    i--;
        }
        return stack.stream().mapToInt(i -> i).toArray();
    }
}

File: JavaScript/Medium/AsteroidCollision.js
/**
 * 735. Asteroid Collision
 * @param {number[]} asteroids
 * @return {number[]}
 */
var asteroidCollision = function (asteroids) {
  const res = [];
  for (let i = 0; i < asteroids.length; i++) {
    if (asteroids[i] > 0 || res.length === 0 || res[res.length - 1] < 0) {
      res.push(asteroids[i]);
    } else if (res[res.length - 1] <= -asteroids[i]) {
      if (res.pop() < -asteroids[i]) {
        i--;
      }
    }
  }
  return res;
};
